src/torch_ucc.cpp

```diff
@@ -718,6 +742,113 @@ ProcessGroupUCC::~ProcessGroupUCC() {
   }
 }
 
+#ifdef USE_CUDA
+// Return CUDA device with ordinal given by input rank.
+at::Device getCUDADeviceForRank(int rank) {
+  TORCH_CHECK(rank >= 0, "Invalid rank ", rank);
+  auto numGPUs = at::cuda::getNumGPUs();
```

Comment:
I think we need to handle the case when running CUDA-enable Pytorch build on non-GPU nodes (`at::cuda::getNumGPUs() ==0`). Perhaps just repeats `c10::kCPU` (or some dummy value we can skip healthcheck later) for such corner case?

Reply:
I changed the code in `runHealthCheck` to
```C++
    std::list<c10::Device> devices{c10::kCPU};
    if (at::cuda::is_available()) {
      devices.emplace_front(getCUDADeviceForRank(rank_));
    }
    for (auto device : devices) {
```

Reply:
I think `at::cuda::is_available()` won't guarantee `at::cuda::getNumGPUs() > 0`. Say a user runs the same CUDA-enabled pytorch build on non-GPU nodes (e.g., all CUDA libraries are installed, but no physical GPU), or a strange but valid scenario: a user runs with CUDA_VISIBLE_DEVICES="" on GPU nodes. In such case, I believe `at::cuda::is_available()` will return `true`, but `at::cuda::getNumGPUs()` return `0`, which will likely lead to unexpected errors later.

should we do the check like this?
```
    if (at::cuda::is_available() && at::cuda::getNumGPUs() > 0) {
      devices.emplace_front(getCUDADeviceForRank(rank_));
    }
```

Reply:
checking the number of GPUs is exactly what `is_available` does:
https://github.com/pytorch/pytorch/blob/b08309ee0a3b3b8f3fbe6ef469737378f864208a/aten/src/ATen/cuda/CUDAContext.h#L55
```C++
/**
 * CUDA is available if we compiled with CUDA, and there are one or more
 * devices.  If we compiled with CUDA but there is a driver problem, etc.,
 * this function will report CUDA is not available (rather than raise an error.)
 */
inline bool is_available() {
    return c10::cuda::device_count() > 0;
}

```

Reply:
Thanks for the clarification and the pointer!